apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DatabaseManagerImpl.java

```diff
@@ -185,23 +215,41 @@ public TransactionalDataSource createAndAddShard(Long shardId, DbVersion dbVersi
     }
 
     @Override
-    public synchronized List<TransactionalDataSource> getFullDataSources() {
+    public synchronized List<TransactionalDataSource> getFullDataSources(Long numberOfShards) {
         Set<Long> allFullShards = findAllFullShardId();
-        List<TransactionalDataSource> dataSources = allFullShards.stream().sorted(
-                Comparator.reverseOrder()).map(id-> getOrCreateShardDataSourceById(
-                        id, new ShardAddConstraintsSchemaVersion())).collect(Collectors.toList());
+        List<TransactionalDataSource> dataSources;
+        if (numberOfShards != null) {
+            dataSources = allFullShards.stream().limit(numberOfShards).sorted(
+                    Comparator.reverseOrder()).map(id -> getOrCreateShardDataSourceById(
+                    id, new ShardAddConstraintsSchemaVersion())).collect(Collectors.toList());
+        } else {
+            dataSources = allFullShards.stream().sorted(
+                    Comparator.reverseOrder()).map(id -> getOrCreateShardDataSourceById(
+                    id, new ShardAddConstraintsSchemaVersion())).collect(Collectors.toList());
+        }
         return dataSources;
     }
 
     @Override
-    public synchronized int closeAllShardDataSources() {
-        int closedDatasources = 0;
-        for (TransactionalDataSource dataSource : connectedShardDataSourceMap.values()) {
+    public Iterator<TransactionalDataSource> getFullDataSourcesIterator() {
```

Comment:
the same as above

Reply:
method name is renamed, java doc is added